src/socket_base.cpp

```diff
@@ -1580,6 +1587,9 @@ int zmq::socket_base_t::monitor (const char *addr_, int events_)
         errno = ETERM;
         return -1;
     }
+
+    scoped_lock_t lock(monitor_sync);
```

Comment:
Won't having the lock after the check for ctx_terminated risk another race condition? In case process_stop and monitor are called from 2 different threads